src/txdb.cpp

```diff
@@ -424,3 +424,132 @@ bool CCoinsViewDB::Upgrade() {
+        batch.Write(std::make_pair(DB_TXINDEX, tuple.first), tuple.second);
+    }
+    return WriteBatch(batch);
+}
+
+bool TxIndexDB::ReadBestBlockHash(uint256& hash) const {
+    if (Read(DB_BEST_BLOCK, hash)) {
+        return true;
+    }
+
+    // Read might have failed either because key does not exist or due to an error.
+    // If the former, return value should still be true.
+    if (!Exists(DB_BEST_BLOCK)) {
+        hash.SetNull();
+        return true;
+    }
+    return false;
+}
+
+bool TxIndexDB::WriteBestBlockHash(const uint256& hash) {
+    return Write(DB_BEST_BLOCK, hash);
+}
+
+bool TxIndexDB::MigrateData(CBlockTreeDB& block_tree_db, const uint256& block_hash)
+{
+    // The prior implementation of txindex was always in sync with block index
+    // and presence was indicated with a boolean DB flag. If the flag is not
+    // set, txindex did not exist or was already migrated.
+    bool f_migrate_index = false;
+    block_tree_db.ReadFlag("txindex", f_migrate_index);
+    if (!f_migrate_index) {
+        return true;
+    }
+
+    int64_t count = 0;
+    LogPrintf("Upgrading txindex database...\n");
+    LogPrintf("[0%%]...");
+    uiInterface.ShowProgress(_("Upgrading txindex database"), 0, true);
+    int report_done = 0;
+    const size_t batch_size = 1 << 24; // 16 MiB
+
+    CDBBatch batch_newdb(*this);
+    CDBBatch batch_olddb(block_tree_db);
+
+    std::pair<unsigned char, uint256> key;
+    std::pair<unsigned char, uint256> begin_key{DB_TXINDEX, uint256()};
+    std::pair<unsigned char, uint256> prev_key_newdb = begin_key;
+    std::pair<unsigned char, uint256> prev_key_olddb = begin_key;
+
+    std::unique_ptr<CDBIterator> pcursor(block_tree_db.NewIterator());
+    for (pcursor->Seek(begin_key); pcursor->Valid(); pcursor->Next()) {
+        boost::this_thread::interruption_point();
+        if (ShutdownRequested()) {
+            break;
+        }
+
+        if (!pcursor->GetKey(key)) {
+            return error("%s: cannot get key from valid cursor", __func__);
+        }
+        if (key.first != DB_TXINDEX) {
+            break;
+        }
+
+        // Log progress every 10%.
+        if (++count % 256 == 0) {
+            uint32_t high = 0x100 * *key.second.begin() + *(key.second.begin() + 1);
+            int percentage_done = (int)(high * 100.0 / 65536.0 + 0.5);
+            uiInterface.ShowProgress(_("Upgrading txindex database"), percentage_done, true);
+            if (report_done < percentage_done/10) {
+                LogPrintf("[%d%%]...", percentage_done);
+                report_done = percentage_done/10;
+            }
+        }
+
+        CDiskTxPos value;
+        if (!pcursor->GetValue(value)) {
+            return error("%s: cannot parse txindex record", __func__);
+        }
+        batch_newdb.Write(key, value);
+        batch_olddb.Erase(key);
+
+        if (batch_newdb.SizeEstimate() > batch_size || batch_olddb.SizeEstimate() > batch_size) {
+            WriteBatch(batch_newdb);
```

Comment:
The transaction logic throughout is wrong here because there's a data ordering dependency.

During the data migration:
 * Write to new new with `fsync = True`
 * After new write finishes, you can erase from old db (fsync setting can be left false)

This ensures that you can can always scan through the old database and catch up to what's in the new database if there's a crash or you are interrupted. During the non-migration path I think it's fine to have fsync disabled.